File: ulib/ml/FStar_String.ml
let make i ci = String.make (Z.to_int i) (Char.chr ci)
let strcat s t = s ^ t
let split seps s =
  let rec repeat_split acc = function
    | [] -> acc
    | sep::seps ->
       let usep = BatUTF8.init 1 (fun _ -> BatUChar.chr sep) in
       let l = BatList.flatten (BatList.map (fun x -> BatString.nsplit x usep) acc) in
       repeat_split l seps in
  repeat_split [s] seps
let compare x y = Z.of_int (BatString.compare x y)
let concat = BatString.concat
let length s = Z.of_int (BatUTF8.length s)
let strlen s = length s

let substring s i j =
  BatUTF8.init (Z.to_int j) (fun k -> BatUTF8.get s (k + Z.to_int i))
let sub s i j =
   substring s i (Z.of_int (BatUTF8.length s - (Z.to_int j) + 1))

let get s i = BatUChar.code (BatUTF8.get s (Z.to_int i))
let collect f s =
  let r = ref "" in
  BatUTF8.iter (fun c -> r := !r ^ f (BatUChar.code c)) s; !r
let lowercase = BatString.lowercase
let uppercase = BatString.uppercase
let index = get
let sub = substring
let list_of_string s = BatList.init (BatUTF8.length s) (fun i -> BatUChar.code (BatUTF8.get s i))
let string_of_list l = BatUTF8.init (BatList.length l) (fun i -> BatUChar.chr (BatList.at l i))

File: stage0/ulib/fs/FStar_String.fs
module FStar_String
open Prims

let make (i : nat) (c : FStar_Char.char) = String.init (Microsoft.FSharp.Core.Operators.int i) (fun _ -> string([|c|]))
let strcat s t = s ^ t
let op_Hat s t =  strcat s t

let split (seps : FStar_Char.char list) (s : string) = s.Split(Array.ofList seps)
  
let compare (x : string) (y : string) = Prims.of_int (x.CompareTo(y)) 
type char = FStar_Char.char
let concat = String.concat
let length s = Prims.of_int (String.length s)
let strlen s = length s

let substring (s : string) (i : Prims.int) (j : Prims.int) = s.Substring(Microsoft.FSharp.Core.Operators.int i, Microsoft.FSharp.Core.Operators.int j)
let sub = substring

let get (s : string) (i : Prims.int) = s.[Microsoft.FSharp.Core.Operators.int i]
let collect (f : char -> string) (s : string) = s |> Array.ofSeq |> Array.map f |> String.concat ""
let lowercase (s : string) = s.ToLowerInvariant()
let uppercase (s : string) = s.ToUpperInvariant()
//let escaped = BatString.escaped
let index = get

let index_of (s : string) (c : char) = s.IndexOf(c) 
let list_of_string (s : string) = s |> Seq.toList 
let string_of_list (l : char list) = string(Array.ofList l)
let string_of_char (c : char) =  string(c, 1)
